pkg/service/application_cache.go

```diff
@@ -0,0 +1,70 @@
+package service
+
+import (
+	"context"
+	"reflect"
+	"time"
+
+	"github.com/pyroscope-io/pyroscope/pkg/storage"
+)
+
+type ApplicationWriter interface {
+	CreateOrUpdate(ctx context.Context, application storage.Application) error
+}
+
+type ApplicationCacheService struct {
+	appSvc ApplicationWriter
+	cache  *cache
+}
+
+type ApplicationCacheServiceConfig struct {
+	Size int
+	TTL  time.Duration
+}
+
+func NewApplicationCacheService(config ApplicationCacheServiceConfig, appSvc ApplicationWriter) *ApplicationCacheService {
+	if config.Size <= 0 {
+		config.Size = 100
+	}
+
+	if config.TTL <= 0 {
+		config.TTL = 5 * time.Minute
+	}
+
+	cache := newCache(config.Size, config.TTL)
+	return &ApplicationCacheService{appSvc: appSvc, cache: cache}
+}
+
+// CreateOrUpdate delegates to the underlying service in the following cases:
+// * item is not in the cache
+// * data is different from what's in the cache
+// Otherwise it does nothing
+func (svc *ApplicationCacheService) CreateOrUpdate(ctx context.Context, application storage.Application) error {
+	if cachedApp, ok := svc.cache.get(application.FullyQualifiedName); ok {
+		if !svc.isTheSame(application, cachedApp.(storage.Application)) {
+			return svc.writeToBoth(ctx, application)
+		}
+		return nil
+	}
+
+	// Not in cache
+	// Could be due to TTL
+	// Or could it be that's a new app
+	return svc.writeToBoth(ctx, application)
+}
```

Comment:
I would maybe ensure that updates are throttled, e.g. once per second / 10 seconds. There is one problem that agents _may_ send different data :/ `SampleRate` causes most of concerns. I would maybe consider ignoring it and eventually removed it from the application metadata